### PR TITLE
GEOMESA-2683 Converter Metrics - AWS CloudWatch

### PR DESCRIPTION
**Pull request by forrestfeaser**
Signed-off-by: Forrest Feaser <forrest.feaser@ccri.com>

### CODE REVIEW

File: build/cqs.tsv [hunk 1]
```diff
@@ -150,6 +156,8 @@ org.apache.hbase:hbase-protocol	1.3.1	compile
 org.apache.hbase:hbase-server	1.3.1	compile
 org.apache.hive:hive-storage-api	2.6.0	compile
 org.apache.htrace:htrace-core	3.1.0-incubating	compile
+org.apache.httpcomponents:httpclient	4.5.3	compile
+org.apache.httpcomponents:httpcore	4.4.6	compile
```
Comment (jnh5y) [inline build/cqs.tsv]: Hmm... it is very odd that these versions do not match.

Comment (forrestfeaser) [reply build/cqs.tsv]: httpcore only goes up to 4.4.11, but I could try lowering the httpclient version and see if it still works

File: build/cqs.tsv [hunk 2]
```diff
@@ -11,6 +14,7 @@ com.esotericsoftware:minlog	1.3.0	compile
 com.fasterxml.jackson.core:jackson-annotations	2.9.7	compile
 com.fasterxml.jackson.core:jackson-core	2.9.7	compile
 com.fasterxml.jackson.core:jackson-databind	2.9.7	compile
+com.fasterxml.jackson.dataformat:jackson-dataformat-cbor	2.6.7	compile
```
Comment (elahrvivaz) [inline build/cqs.tsv]: we should add a pom entry to make this match the rest of the jackson jar versions, at 2.9.7. Those are defined here: https://github.com/locationtech/geomesa/blob/master/pom.xml#L1723

File: build/cqs.tsv [hunk 3]
```diff
@@ -76,6 +80,8 @@ io.airlift:aircompressor	0.10	compile
 io.dropwizard.metrics:metrics-core	3.2.6	compile
 io.dropwizard.metrics:metrics-ganglia	3.2.6	compile
 io.dropwizard.metrics:metrics-graphite	3.2.6	compile
+io.dropwizard.metrics:metrics-jvm	4.0.2	compile
```
Comment (elahrvivaz) [inline build/cqs.tsv]: this is going to need to match the rest of the metrics versions too. @jahhulbert-ccri just asked us to bump the metrics version - looks like 4.1.0 is the latest. Could you try setting the version to that, and see if things still build ok? We'll also need to add an entry for metrics-jvm here: https://github.com/locationtech/geomesa/blob/master/pom.xml#L1584

File: build/cqs.tsv (hunk 1 above)
Comment (elahrvivaz) [reply build/cqs.tsv]: from looking at the poms, they don't always match. For instance, httpclient 4.5.9 (the latest) depends on httpcore 4.4.11

File: build/cqs.tsv (hunk 3 above)
Comment (forrestfeaser) [reply build/cqs.tsv]: When I changed the metrics version I get the following error during the build: Could not resolve dependencies for project org.locationtech.geomesa:geomesa-metrics_2.11:jar:2.4.0-SNAPSHOT: Failure to find io.dropwizard.metrics:metrics-ganglia:jar:4.1.0 in http://art.ccri.com/artifactory/libs-release was cached in the local repository, resolution will not be reattempted until the update interval of libs-release has elapsed or updates are forced

Comment (elahrvivaz) [reply build/cqs.tsv]: ok, thanks - to not hold up this PR, can you just set the metrics-jvm version to 3.2.6, and we can update everything to 4.x later.

File: build/cqs.tsv (hunk 1 above)
Comment (elahrvivaz) [reply build/cqs.tsv]: do we actually need httpclient? It's not mentioned in the reporter libs, unless they are transitive deps: https://github.com/azagniotov/codahale-aggregated-metrics-cloudwatch-reporter#third-party-dependencies. But I notice they are explicitly declared in our pom.

Comment (forrestfeaser) [reply build/cqs.tsv]: I can try taking it out and seeing if it works

Comment (forrestfeaser) [reply build/cqs.tsv]: It did not work without httpclient 4.5.3, but I pushed the other changes back up

File: docs/user/convert/parsing_and_validation.rst [hunk 4]
```diff
@@ -191,17 +191,26 @@ through the converter evaluation context, or can be exposed through reporters co
...
     geomesa.converters.myconverter {
       options {
         reporters = [
+          {
+            type        = "cloudwatch"
+            units       = "MILLISECONDS"
+            interval    = "60 seconds"
+            namespace   = "geomesa"
+            raw-counts  = false
+            zero-values = true
```
Comment (jnh5y) [inline docs/user/convert/parsing_and_validation.rst]: How does the code authenticate with AWS CloudWatch to post metrics?  How is any of that configured?

Comment (forrestfeaser) [reply docs/user/convert/parsing_and_validation.rst]: It uses your default credentials and region. Should I make that configurable?

Comment (jnh5y) [reply docs/user/convert/parsing_and_validation.rst]: What do you mean by 'default credentials'?  Where do those defaults come from?

Comment (forrestfeaser) [reply docs/user/convert/parsing_and_validation.rst]: i think .aws/credentials

Comment (jnh5y) [reply docs/user/convert/parsing_and_validation.rst]: Hmm... there can be profiles in there...  How does those interact?  Also, if we are pulling from default locations, it is worth documenting that.

Comment (forrestfeaser) [reply docs/user/convert/parsing_and_validation.rst]: One of the credentials in there should be listed as default, and I could make it so you have to configure your credentials in the converter if that is preferable.

File: build/cqs.tsv [hunk 5]
```diff
@@ -1,5 +1,8 @@
 asm:asm	3.1	compile
 com.adobe.xmp:xmpcore	5.1.2	compile
+com.amazonaws:aws-java-sdk-cloudwatch	1.11.179	compile
+com.amazonaws:aws-java-sdk-core	1.11.179	compile
+com.amazonaws:jmespath-java	1.11.179	compile
```
Comment (elahrvivaz) [inline build/cqs.tsv]: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=20589

File: build/cqs.tsv [hunk 6]
```diff
@@ -11,6 +14,7 @@ com.esotericsoftware:minlog	1.3.0	compile
 com.fasterxml.jackson.core:jackson-annotations	2.9.7	compile
 com.fasterxml.jackson.core:jackson-core	2.9.7	compile
 com.fasterxml.jackson.core:jackson-databind	2.9.7	compile
+com.fasterxml.jackson.dataformat:jackson-dataformat-cbor	2.9.7	compile
```
Comment (elahrvivaz) [inline build/cqs.tsv]: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=20590

File: build/cqs.tsv [hunk 7]
```diff
@@ -76,6 +80,8 @@ io.airlift:aircompressor	0.10	compile
 io.dropwizard.metrics:metrics-core	3.2.6	compile
 io.dropwizard.metrics:metrics-ganglia	3.2.6	compile
 io.dropwizard.metrics:metrics-graphite	3.2.6	compile
+io.dropwizard.metrics:metrics-jvm	3.2.6	compile
```
Comment (elahrvivaz) [inline build/cqs.tsv]: @jnh5y I am going to skip opening a CQ for this, as we're planning to bump metrics to 4.x. If we don't bump it for whatever reason, don't let me forget to CQ this one before release.

File: build/cqs.tsv [hunk 8]
```diff
@@ -246,6 +254,7 @@ picocontainer:picocontainer	1.2	compile
 redis.clients:jedis	3.0.1	compile
 si.uom:si-quantity	0.7.1	compile
 si.uom:si-units-java8	0.7.1	compile
+software.amazon.ion:ion-java	1.0.2	compile
```
Comment (elahrvivaz) [inline build/cqs.tsv]: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=20591

File: build/cqs.tsv [hunk 9]
```diff
@@ -76,6 +80,8 @@ io.airlift:aircompressor	0.10	compile
 io.dropwizard.metrics:metrics-core	3.2.6	compile
 io.dropwizard.metrics:metrics-ganglia	3.2.6	compile
 io.dropwizard.metrics:metrics-graphite	3.2.6	compile
+io.dropwizard.metrics:metrics-jvm	3.2.6	compile
+io.github.azagniotov:dropwizard-metrics-cloudwatch	1.0.13	compile
```
Comment (elahrvivaz) [inline build/cqs.tsv]: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=20592

File: build/cqs.tsv [hunk 10]
```diff
@@ -150,6 +156,8 @@ org.apache.hbase:hbase-protocol	1.3.1	compile
 org.apache.hbase:hbase-server	1.3.1	compile
 org.apache.hive:hive-storage-api	2.6.0	compile
 org.apache.htrace:htrace-core	3.1.0-incubating	compile
+org.apache.httpcomponents:httpclient	4.5.3	compile
```
Comment (elahrvivaz) [inline build/cqs.tsv]: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=20593

File: build/cqs.tsv (hunk 1 above)
Comment (elahrvivaz) [inline build/cqs.tsv]: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=20594

File: build/cqs.tsv (hunk 8 above)
Comment (elahrvivaz) [reply build/cqs.tsv]: approved

File: build/cqs.tsv (hunk 10 above)
Comment (elahrvivaz) [reply build/cqs.tsv]: approved

File: build/cqs.tsv (hunk 1 above)
Comment (elahrvivaz) [reply build/cqs.tsv]: approved

File: build/cqs.tsv (hunk 5 above)
Comment (jnh5y) [reply build/cqs.tsv]: Approved!

Comment (elahrvivaz) [reply build/cqs.tsv]: approved

File: build/cqs.tsv (hunk 6 above)
Comment (elahrvivaz) [reply build/cqs.tsv]: approved

File: build/cqs.tsv (hunk 9 above)
Comment (elahrvivaz) [reply build/cqs.tsv]: approved

**Pull request merged**